src/coreclr/jit/gentree.cpp

```diff
@@ -17652,6 +17658,197 @@ bool GenTreeVecCon::HandleArgForHWIntrinsicCreate(GenTree* arg, int argIdx, simd
 
     return false;
 }
+
+// TODO-XArch-AVX512: Keep only one implementation once GenTreeVecCon supports gtSimd64Val.
+#if defined(TARGET_XARCH)
+//----------------------------------------------------------------------------------------------
+// IsHWIntrinsicCreateConstant: Determines if a HWIntrinsic node represents a vector constant
+//
+//  Arguments:
+//     node      - The node to check
+//     simd64Val - The vector constant being constructed
+//
+//  Returns:
+//     true if node represents a constant; otherwise, false
+bool GenTreeVecCon::IsHWIntrinsicCreateConstant(GenTreeHWIntrinsic* node, simd64_t& simd64Val)
+{
+    NamedIntrinsic intrinsic    = node->GetHWIntrinsicId();
+    var_types      simdType     = node->TypeGet();
+    var_types      simdBaseType = node->GetSimdBaseType();
+    unsigned       simdSize     = node->GetSimdSize();
+
+    size_t argCnt    = node->GetOperandCount();
+    size_t cnsArgCnt = 0;
+
+    switch (intrinsic)
+    {
+        case NI_Vector512_Create:
+        {
+            // Zero out the simd64Val
+            simd64Val = {};
+
+            // These intrinsics are meant to set the same value to every element.
+            if ((argCnt == 1) && HandleArgForHWIntrinsicCreate(node->Op(1), 0, simd64Val, simdBaseType))
+            {
+                cnsArgCnt = 1;
+            }
+            else
+            {
+                for (unsigned i = 1; i <= argCnt; i++)
+                {
+                    if (HandleArgForHWIntrinsicCreate(node->Op(i), i - 1, simd64Val, simdBaseType))
+                    {
+                        cnsArgCnt++;
+                    }
+                }
+            }
+
+            assert((argCnt == 1) || (argCnt == (simdSize / genTypeSize(simdBaseType))));
+            return argCnt == cnsArgCnt;
+        }
+
+        default:
+        {
+            return false;
+        }
+    }
+}
+
+//----------------------------------------------------------------------------------------------
+// HandleArgForHWIntrinsicCreate: Processes an argument for the GenTreeVecCon::IsHWIntrinsicCreateConstant method
+//
+//  Arguments:
+//     arg       - The argument to process
+//     argIdx    - The index of the argument being processed
+//     simd32Val - The vector constant being constructed
+//     baseType  - The base type of the vector constant
+//
+//  Returns:
+//     true if arg was a constant; otherwise, false
+bool GenTreeVecCon::HandleArgForHWIntrinsicCreate(GenTree* arg, int argIdx, simd64_t& simd64Val, var_types baseType)
```

Comment:
There is quite a bit of code duplication here that could be removed if the method was templated or passed in a pointer to the `simdVal` instead.

Then we could have one function that worked with any of `simd16/32/64_t`.

Reply:
-- This doesn't necessarily mean that the `simd32_t&` and `simd64_t&` overloads needs to be removed, if that will cause too much unnecessary churn.

But rather, they could both call into an overload that takes `void*` and has the shared impl logic.

Reply:
Right. I switched to a templated function - 
```
template <typename simdTypename>
    static bool IsHWIntrinsicCreateConstant(GenTreeHWIntrinsic* node, simdTypename& simdVal)
```    
    Had to do some moving around
    1. Move definition to header to deal with link time ymbol resolution issues
    2. Move location of `GenTreeVecConGenTreeVecCon `below `GenTreeHWIntrinsic`
    3. Add a helper to keep `genTypeSize()` in cpp file(easier that way than try to include certain headers)